ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/classic/JmsXAResourceRecoveryHelper.java

```diff
@@ -0,0 +1,182 @@
+            LOGGER.trace("Initialise with properties=" + properties);
+        }
+
+        return true;
+    }
+
+    @Override
+    public XAResource[] getXAResources() {
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace("Returning XA resource: " + this);
+        }
+
+        return new XAResource[] { this };
+    }
+
+    @Override
+    public Xid[] recover(int i) throws XAException {
+        if (i == XAResource.TMSTARTRSCAN) {
+            connect();
+        }
+
+        assert delegate != null : "Recovery scan has to be started";
+
+        try {
+            return delegate.recover(i);
+        } finally {
+            if (i == XAResource.TMENDRSCAN) {
+                disconnect();
+            }
+        }
+    }
+
+    @Override
+    public void start(Xid xid, int i) throws XAException {
+        assert delegate != null : "Recovery scan has to be started";
+
+        delegate.start(xid, i);
+    }
+
+    @Override
+    public void end(Xid xid, int i) throws XAException {
+        assert delegate != null : "Recovery scan has to be started";
+
+        delegate.end(xid, i);
+    }
+
+    @Override
+    public int prepare(Xid xid) throws XAException {
+        assert delegate != null : "Recovery scan has to be started";
+
+        return delegate.prepare(xid);
+    }
+
+    @Override
+    public void commit(Xid xid, boolean b) throws XAException {
+        assert delegate != null : "Recovery scan has to be started";
+
+        delegate.commit(xid, b);
+    }
+
+    @Override
+    public void rollback(Xid xid) throws XAException {
+        assert delegate != null : "Recovery scan has to be started";
+
+        delegate.rollback(xid);
+    }
+
+    @Override
+    public boolean isSameRM(XAResource xaResource) throws XAException {
+        assert delegate != null : "Recovery scan has to be started";
+
+        return delegate.isSameRM(xaResource);
+    }
+
+    @Override
+    public void forget(Xid xid) throws XAException {
+        assert delegate != null : "Recovery scan has to be started";
+
+        delegate.forget(xid);
+    }
+
+    @Override
+    public int getTransactionTimeout() throws XAException {
+        assert delegate != null : "Recovery scan has to be started";
+
+        return delegate.getTransactionTimeout();
+    }
+
+    @Override
+    public boolean setTransactionTimeout(int i) throws XAException {
+        assert delegate != null : "Recovery scan has to be started";
+
+        return delegate.setTransactionTimeout(i);
+    }
+
+    private void connect() throws XAException {
+        if (delegate != null) {
+            return;
+        }
+
+        try {
+            xaConnection = xaConnectionFactory.createXAConnection();
+            xaSession = xaConnection.createXASession();
+            delegate = xaSession.getXAResource();
+        } catch (JMSException e) {
+            LOGGER.warn("Failed to create connection: " + e.getMessage());
+            throw new XAException(e.getMessage());
```

Comment:
Is it possible to set the e as the root cause on the thrown XAException so you don't lose the stack trace